scripts/1.TCGA-MLexample.py

```diff
@@ -72,84 +75,96 @@
         urllib.request.urlretrieve(url, path)
 
 
-# In[7]:
+# In[8]:
 
 get_ipython().run_cell_magic('time', '', "path = os.path.join('data', 'expression.tsv.bz2')\nX = pd.read_table(path, index_col=0)")
 
 
-# In[8]:
+# In[9]:
 
 get_ipython().run_cell_magic('time', '', "path = os.path.join('data', 'mutation-matrix.tsv.bz2')\nY = pd.read_table(path, index_col=0)")
 
 
-# ## Process Mutation Matrix
-
-# In[9]:
+# In[10]:
 
 y = Y[GENE]
 
 
-# In[10]:
+# In[11]:
 
 # The Series now holds TP53 Mutation Status for each Sample
 y.head(6)
 
 
-# In[11]:
+# In[12]:
 
 # Here are the percentage of tumors with NF1
 y.value_counts(True)
 
 
 # ## Set aside 10% of the data for testing
 
-# In[12]:
+# In[13]:
 
 # Typically, this can only be done where the number of mutations is large enough
 X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
 'Size: {:,} features, {:,} training samples, {:,} testing samples'.format(len(X.index), len(X_train), len(X_test))
 
 
-# ## Feature transformation: Standardization
+# ## Median absolute deviation feature selection
 
-# In[13]:
+# In[14]:
 
-def scale_df(df):
-    """Standardize colummns by scaling to z-scores"""
-    scaled_matrix = preprocessing.scale(df)
-    return pd.DataFrame(scaled_matrix, index=df.index, columns=df.columns)
+def fs_mad(x, y):
+    mad = lambda a: np.median(np.absolute(a - np.median(a)))
+    score = np.apply_along_axis(mad, 0, x)
+    p = np.array([0.5]*len(score))
+    return score, p
+feature_select = SelectKBest(fs_mad, k=500)  # select the top 500 features with the highest MAD
 
-# Standardize X separately for the holdout dataset
-X_train, X_test = scale_df(X_train), scale_df(X_test)
 
+# ## Define pipeline and Cross validation model fitting
 
-# ## Cross validation model fitting
+# In[15]:
 
-# In[14]:
+# Include loss='log' in param_grid doesn't work with pipeline somehow
+clf = SGDClassifier(random_state=0, class_weight='balanced', loss='log')
```

Comment:
Thanks for your comments. I agreed with most of them. For the param_grid though, it seems that if loss is in param_grid as in the following code
```python
param_grid = {
    'loss': ['log'],
    'penalty': ['elasticnet'],
    'alpha': [10 ** x for x in range(-6, 1)],
    'l1_ratio': [0, 0.05, 0.1, 0.2, 0.5, 0.8, 0.9, 0.95, 1],
}
clf = SGDClassifier(random_state=0, class_weight='balanced')
clf_grid = grid_search.GridSearchCV(estimator=clf, param_grid=param_grid, n_jobs=-1, scoring='roc_auc')
pipeline = make_pipeline(
              feature_select,  # Feature selection
              StandardScaler(),  # Feature scaling
              clf_grid)
pipeline.fit(X=X_train, y=y_train)
pipeline.predict_proba(X)
```
I will get the following error message
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-28-f384560df3ec> in <module>()
----> 1 pipeline.predict_proba(X)

/home/yichuanliu/Programs/anaconda3/envs/cognoma-cancer-data/lib/python3.5/site-packages/sklearn/utils/metaestimators.py in __get__(self, obj, type)
     33             # delegate only on instances, not the classes.
     34             # this is to allow access to the docstrings.
---> 35             self.get_attribute(obj)
     36         # lambda, but not partial, allows help() to work with update_wrapper
     37         out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)

/home/yichuanliu/Programs/anaconda3/envs/cognoma-cancer-data/lib/python3.5/site-packages/sklearn/utils/metaestimators.py in __get__(self, obj, type)
     33             # delegate only on instances, not the classes.
     34             # this is to allow access to the docstrings.
---> 35             self.get_attribute(obj)
     36         # lambda, but not partial, allows help() to work with update_wrapper
     37         out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)

/home/yichuanliu/Programs/anaconda3/envs/cognoma-cancer-data/lib/python3.5/site-packages/sklearn/linear_model/stochastic_gradient.py in predict_proba(self)
    754         http://jmlr.csail.mit.edu/papers/volume2/zhang02c/zhang02c.pdf
    755         """
--> 756         self._check_proba()
    757         return self._predict_proba
    758 

/home/yichuanliu/Programs/anaconda3/envs/cognoma-cancer-data/lib/python3.5/site-packages/sklearn/linear_model/stochastic_gradient.py in _check_proba(self)
    719         if self.loss not in ("log", "modified_huber"):
    720             raise AttributeError("probability estimates are not available for"
--> 721                                  " loss=%r" % self.loss)
    722 
    723     @property

AttributeError: probability estimates are not available for loss='hinge'
```

My guess is this is probably a bug in grid_search. I think one solution is to specify the loss  and other parameters that do not need to be grid searched outside of param_grid:
```
# Parameter Sweep for Hyperparameters
SGD_loss = 'log'
SGD_penalty = 'elasticnet'
n_feature_kept = 500
param_grid = {
    'alpha': [10 ** x for x in range(-6, 1)],
    'l1_ratio': [0, 0.05, 0.1, 0.2, 0.5, 0.8, 0.9, 0.95, 1],
}
feature_select = SelectKBest(fs_mad, k=n_feature_kept)
clf = SGDClassifier(random_state=0, class_weight='balanced', penalty=SGD_penalty, loss=SGD_loss)
clf_grid = grid_search.GridSearchCV(estimator=clf, param_grid=param_grid, n_jobs=-1, scoring='roc_auc')
pipeline = make_pipeline(
              feature_select,  # Feature selection
              StandardScaler(),  # Feature scaling
              clf_grid)
```
Let me know your opionion